bigquery/src/main/java/org/apache/iceberg/gcp/bigquery/BigQueryMetastoreCatalog.java

```diff
@@ -0,0 +1,394 @@
+
+    String fileIoImpl =
+        properties.getOrDefault(
+            CatalogProperties.FILE_IO_IMPL, "org.apache.iceberg.hadoop.HadoopFileIO");
+    this.fileIO = CatalogUtil.loadFileIO(fileIoImpl, properties, conf);
+
+    this.listAllTables = Boolean.parseBoolean(properties.getOrDefault(LIST_ALL_TABLES, "true"));
+  }
+
+  @Override
+  protected TableOperations newTableOps(TableIdentifier identifier) {
+    return new BigQueryTableOperations(client, fileIO, toTableReference(identifier), conf);
+  }
+
+  @Override
+  protected String defaultWarehouseLocation(TableIdentifier identifier) {
+    String locationUri = null;
+    DatasetReference datasetReference = toDatasetReference(identifier.namespace());
+    Dataset dataset = client.load(datasetReference);
+    if (dataset != null && dataset.getExternalCatalogDatasetOptions() != null) {
+      locationUri = dataset.getExternalCatalogDatasetOptions().getDefaultStorageLocationUri();
+    }
+
+    return String.format(
+        "%s/%s",
+        Strings.isNullOrEmpty(locationUri)
+            ? createDefaultStorageLocationUri(datasetReference.getDatasetId())
+            : LocationUtil.stripTrailingSlash(locationUri),
+        identifier.name());
+  }
+
+  @Override
+  public List<TableIdentifier> listTables(Namespace namespace) {
+    validateNamespace(namespace);
+
+    return client.list(toDatasetReference(namespace), listAllTables).stream()
+        .map(
+            table -> TableIdentifier.of(namespace.level(0), table.getTableReference().getTableId()))
+        .collect(ImmutableList.toImmutableList());
+  }
+
+  @Override
+  public boolean dropTable(TableIdentifier identifier, boolean purge) {
+    try {
+      TableOperations ops = newTableOps(identifier);
+      TableMetadata lastMetadata = ops.current();
+
+      client.delete(toTableReference(identifier));
+
+      if (purge && lastMetadata != null) {
+        CatalogUtil.dropTableData(ops.io(), lastMetadata);
+      }
+    } catch (NoSuchTableException e) {
+      return false;
+    }
+
+    return true;
+  }
+
+  @Override
+  public void renameTable(TableIdentifier from, TableIdentifier to) {
+    validateNamespace(from.namespace());
+    validateNamespace(to.namespace());
+
+    if (!namespaceExists(to.namespace())) {
+      throw new NoSuchNamespaceException("Namespace does not exist: %s", to.namespace());
+    }
+
+    if (!namespaceExists(from.namespace())) {
+      throw new NoSuchNamespaceException("Namespace does not exist: %s", from.namespace());
+    }
+
+    // Check if the source table exists
+    if (!tableExists(from)) {
+      throw new NoSuchTableException("Table does not exist: %s", from.name());
+    }
+
+    if (!from.namespace().equals(to.namespace())) {
+      throw new ValidationException("New table name must be in the same namespace");
+    }
+
+    // Check if the destination table already exists
+    if (tableExists(to)) {
+      throw new AlreadyExistsException("Table already exists: %s", to.name());
```

Comment:
I think here and in the other error messages during rename it would be beneficial if the error would have `Cannot rename <from> to <to>` as a prefix